allennlp/modules/span_extractors/locally_normalised_span_extractor.py

```diff
@@ -0,0 +1,96 @@
+
+@SpanExtractor.register("locally_normalised")
+class LocallyNormalisedSpanExtractor(SpanExtractor):
+    """
+    Computes span representations by generating an unnormalized attention score for each
+    word in the document, compute distributions over every span with respect to these
+    scores by normalising the attention scores for words inside the span.
+
+    Given these attention distributions over every span, weight the corresponding vector
+    representations of the words in the span by this distribution, returning a weighted
+    representation of each span.
+
+    Parameters
+    ----------
+    input_dim : ``int``, required.
+        The last dimension of the sequence representations.
+
+    Returns
+    -------
+    attended_text_embeddings : ``torch.FloatTensor``
+        A tensor of shape (batch_size, num_spans, input_dim), which each span representation
+        is formed by locally normalising a global attention over the sequence. The only way
+        in which the attention distribution differs over different spans is in the set of words
+        over which they are normalized.
+    """
+    def __init__(self,
+                 input_dim: int) -> None:
+        super().__init__()
+        self._global_attention = TimeDistributed(torch.nn.Linear(input_dim, 1))
+
+    @overrides
+    def forward(self, # pylint: disable=arguments-differ
+                sequence_tensor: torch.FloatTensor,
+                indicies: torch.LongTensor) -> None:
+        # both of shape (batch_size, num_spans, 1)
+        span_starts, span_ends = indicies.split(1, dim=-1)
+
+        # shape (batch_size, num_spans, 1)
+        span_widths = span_ends - span_starts
+
+        # We need to know the maximum span width so we can
+        # generate indices to extract the spans from the sequence tensor.
+        # These indices will then get masked below, such that if the length
+        # of a given span is smaller than the max, the rest of the values
+        # are masked.
+        max_batch_span_width = int(span_widths.max().data) + 1
+
+        # shape (batch_size, sequence_length, 1)
+        global_attention_logits = self._global_attention(sequence_tensor)
+
+        # Shape: (1, 1, max_batch_span_width)
+        max_span_range_indices = util.get_range_vector(max_batch_span_width,
+                                                       sequence_tensor.is_cuda).view(1, 1, -1)
+        # Shape: (batch_size, num_spans, max_batch_span_width)
+        # This is a broadcasted comparison - for each span we are considering,
+        # we are creating a range vector of size max_span_width, but masking values
+        # which are greater than the actual length of the span.
+        span_mask = (max_span_range_indices <= span_widths).float()
+        raw_span_indices = span_ends - max_span_range_indices
+        # We also don't want to include span indices which are less than zero,
+        # which happens because some spans near the beginning of the sequence
```

Comment:
This comment made me think a whole lot.  I think it's incorrect.  The issue isn't that the span _width_ is smaller than `max_batch_span_width`, it's that the span end _index_ is smaller than `max_batch_span_width`.